tensorpack/tfutils/symbolic_functions.py

```diff
@@ -212,3 +212,149 @@ def saliency_map(output, input, name="saliency_map"):
+
+        loss = tf.reduce_mean(0.5 * (y * match_loss + (1 - y) * missmatch_loss))
+        pos_dist = tf.reduce_mean(y * delta_sqrt, name="pos-dist")
+        neg_dist = tf.reduce_mean((1 - y) * delta_sqrt, name="neg-dist")
+
+        if extra:
+            return loss, pos_dist, neg_dist
+        else:
+            return loss
+
+
+def cosine_loss(left, right, y):
+    """Loss for Siamese networks (cosine version).
+
+    Remarks:
+        Same as `contrastive_loss` but with different similarity measurment.
+
+    Args:
+        left (tf.Tensor): left of image pair
+        right (tf.Tensor): right of image pair
+        y (tf.Tensor): label (1: similar, 0:not similar)
+
+    Returns:
+        tf.Tensor: cosine-loss as scalar (and) average_pos_loss, average_neg_loss
+    """
+
+    def l2_norm(t, eps=1e-12):
+        """return L2 norm for input x
+
+        Args:
+            t (tf.Tensor): input tensor
+            eps (float, optional): constant for numerical stability
+
+        Returns:
+            tf.Tensor: norm of input tensor
+        """
+        with tf.name_scope("l2_norm"):
+            return tf.sqrt(tf.reduce_sum(tf.square(t), 1, True) + eps)
+
+    with tf.name_scope("cosine_loss"):
+        y = 2 * tf.cast(y, tf.float32) - 1
+        pred = tf.reduce_sum(left * right, 1) / (tf.squeeze(l2_norm(left) * l2_norm(right)) + 1e-10)
+
+        return tf.nn.l2_loss(y - pred)
+
+
+def triplet_loss(anchor, positive, negative, margin, extra=False):
+    """Loss for Triplet networks as described in the paper:
+    `Deep Metric Learning using Triplet Network <https://arxiv.org/pdf/1412.6622.pdf>` by Hoffer et al.
+
+    Learn embeddings from an anchor point and a similar input (p) as well as a not similar input (n)
+    Intuitively, a matching-pair (anchor, positive) should have a smaller relative distance
+    than a non-matching pair (anchor, negative).
+
+    .. math::
+        \min \max(0, m + \Vert a-p\Vert^2 - \Vert a-n\Vert^2)
+
+    Args:
+        anchor (tf.Tensor): anchor point
+        positive (tf.Tensor): positiv match
+        negative (tf.Tensor): negativ as missmatch
+        margin (float): horizont for negative examples
+        extra (bool, optional): return additional endpoints
+
+    Returns:
+        tf.Tensor: triplet-loss as scalar (and) average_pos_loss, average_neg_loss
+    """
+
+    with tf.name_scope("triplet_loss"):
+        d_pos = tf.reduce_sum(tf.square(anchor - positive), 1)
+        d_neg = tf.reduce_sum(tf.square(anchor - negative), 1)
+
+        loss = tf.reduce_mean(tf.maximum(0., margin + d_pos - d_neg))
+
+        pos_dist = tf.reduce_mean(tf.sqrt(d_pos + 1e-6))
+        neg_dist = tf.reduce_mean(tf.sqrt(d_neg + 1e-6))
+
+        if extra:
+            return loss, pos_dist, neg_dist
+        else:
+            return loss
+
+
+def soft_triplet_loss(anchor, positive, negative, extra=True):
+    """Loss for triplet networks
+
+    same as Triplet Network Loss but without explicit margin
```

Comment:
Could there be either a formula or a reference to a paper explaining what it actually computes?